## Exercises: Filter and query data

In [1]:
import pandas as pd
autos = pd.read_json("../Data/autos.json")
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   aspiration         205 non-null    object 
 1   body-style         205 non-null    object 
 2   bore               201 non-null    float64
 3   city-mpg           205 non-null    int64  
 4   compression-ratio  205 non-null    float64
 5   curb-weight        205 non-null    int64  
 6   drive-wheels       205 non-null    object 
 7   engine-location    205 non-null    object 
 8   engine-size        205 non-null    int64  
 9   engine-type        205 non-null    object 
 10  fuel-system        205 non-null    object 
 11  fuel-type          205 non-null    object 
 12  height             205 non-null    float64
 13  highway-mpg        205 non-null    int64  
 14  horsepower         203 non-null    float64
 15  length             205 non-null    float64
 16  make               205 non

**Use the "autos" dataset to answer the following questions:**

**Q1:** How many cars from Jaguar are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=3
&nbsp;&nbsp;&nbsp;price=36000.0
</details>

In [2]:
#autos.info()

jaguars = autos[autos["make"] == "jaguar"]
print(len(jaguars))

3


**Q2:** How many cars from Toyota are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=32
&nbsp;&nbsp;&nbsp;price=17669.0
</details>

In [3]:
toyotas = autos[autos["make"] == "toyota"]
most_expensive_toyota = toyotas.sort_values(by="price", ascending=False).head(1)
print(most_expensive_toyota)

    aspiration   body-style  bore  city-mpg  compression-ratio  curb-weight  \
172        std  convertible  3.62        24                9.3         2975   

    drive-wheels engine-location  engine-size engine-type  ...    make  \
172          rwd           front          146         ohc  ...  toyota   

    normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
172             134.0              four           two    4800.0  17669.0   

    stroke  symboling wheel-base width  
172    3.5          2       98.4  65.6  

[1 rows x 26 columns]


**Q3:** What is the length, width and height of the most expensive car in the entire dataset?

<details>
<summary>Answer</summary>
<br>&nbsp;&nbsp;&nbsp;length=199.2
&nbsp;&nbsp;&nbsp;width=72.0
&nbsp;&nbsp;&nbsp;height=55.4
</details>

In [4]:
most_expensive_car = autos.sort_values(by="price", ascending=False).head(1)
print(f"{most_expensive_car["height"].values = }, {most_expensive_car["width"].values = }, {most_expensive_car["length"].values = }")

most_expensive_car["height"].values = array([55.4]), most_expensive_car["width"].values = array([72.]), most_expensive_car["length"].values = array([199.2])


**Q4:** What is the lowest price per horsepower in the dataset, and what brand ("make") is that car?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;price per horsepower=72.84
&nbsp;&nbsp;&nbsp;brand=Toyota
</details>

In [5]:
strongest_cheap_cars = autos.sort_values(by=["horsepower", "price"], ascending=[False, True])
#strongest_cheap_cars[["horsepower", "make", "price"]]

(strongest_cheap_cars["price"] / strongest_cheap_cars["horsepower"]).sort_values(ascending=True).head()

print(f"The lowest costing per horsepower car: (it is a {autos.loc[167]["make"]})")
display(autos.iloc[167])

The lowest costing per horsepower car: (it is a toyota)


aspiration               std
body-style           hardtop
bore                    3.62
city-mpg                  24
compression-ratio        9.3
curb-weight             2540
drive-wheels             rwd
engine-location        front
engine-size              146
engine-type              ohc
fuel-system             mpfi
fuel-type                gas
height                  52.0
highway-mpg               30
horsepower             116.0
length                 176.2
make                  toyota
normalized-losses      134.0
num-of-cylinders        four
num-of-doors             two
peak-rpm              4800.0
price                 8449.0
stroke                   3.5
symboling                  2
wheel-base              98.4
width                   65.6
Name: 167, dtype: object

**Q5:** How many of the cars in the dataset has as many cylinders as they have doors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;cars=95
</details>

In [6]:
autos.query("`num-of-cylinders` == `num-of-doors`")

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
3          std      sedan  3.19        24               10.0         2337   
11         std      sedan  3.50        23                8.8         2395   
20         std      sedan  3.03        38                9.6         1909   
24         std  hatchback  2.97        31                9.4         1967   
25         std      sedan  2.97        31                9.4         1989   
..         ...        ...   ...       ...                ...          ...   
198      turbo      sedan  3.62        17                7.5         3045   
199      turbo      wagon  3.62        17                7.5         3157   
200        std      sedan  3.78        23                9.5         2952   
201      turbo      sedan  3.78        19                8.7         3049   
204      turbo      sedan  3.78        19                9.5         3062   

    drive-wheels engine-location  engine-size engine-type  ...       make  \
3            fwd           front          109         ohc  ...       audi   
11           rwd           front          108         ohc  ...        bmw   
20           fwd           front           90         ohc  ...  chevrolet   
24           fwd           front           90         ohc  ...      dodge   
25           fwd           front           90         ohc  ...      dodge   
..           ...             ...          ...         ...  ...        ...   
198          rwd           front          130         ohc  ...      volvo   
199          rwd           front          130         ohc  ...      volvo   
200          rwd           front          141         ohc  ...      volvo   
201          rwd           front          141         ohc  ...      volvo   
204          rwd           front          141         ohc  ...      volvo   

    normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
3               164.0              four          four    5500.0  13950.0   
11              192.0              four          four    5800.0  16925.0   
20               81.0              four          four    5400.0   6575.0   
24              148.0              four          four    5500.0   6229.0   
25              148.0              four          four    5500.0   6692.0   
..                ...               ...           ...       ...      ...   
198             103.0              four          four    5100.0  18420.0   
199              74.0              four          four    5100.0  18950.0   
200              95.0              four          four    5400.0  16845.0   
201              95.0              four          four    5300.0  19045.0   
204              95.0              four          four    5400.0  22625.0   

    stroke  symboling wheel-base width  
3     3.40          2       99.8  66.2  
11    2.80          0      101.2  64.8  
20    3.11          0       94.5  63.6  
24    3.23          1       93.7  63.8  
25    3.23          1       93.7  63.8  
..     ...        ...        ...   ...  
198   3.15         -2      104.3  67.2  
199   3.15         -1      104.3  67.2  
200   3.15         -1      109.1  68.9  
201   3.15         -1      109.1  68.8  
204   3.15         -1      109.1  68.9  

[95 rows x 26 columns]

**Use the "autos" dataset and write python code to solve the following tasks:**

**T1:** Print a string to inform the user of the price difference between the cheapest and the most expensive car in the dataset.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;The price difference between the cheapest and the most expensive car is 40282.0
</details>

In [7]:
cheapest = autos["price"].sort_values(ascending=True).head(1)
expensivest = autos["price"].sort_values(ascending=False).head(1)
display(cheapest)
display(expensivest)
print(f"The difference of price from the cheapest to the most expensive car is {autos.loc[74]["price"] - autos.loc[138]["price"]} many monies.")

138    5118.0
Name: price, dtype: float64

74    45400.0
Name: price, dtype: float64

The difference of price from the cheapest to the most expensive car is 40282.0 many monies.


**T2:** Ask the user to input a brand, then print the price range for that brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volvo<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'volvo' ranges from 12940.0 to 22625.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: toyota<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'toyota' ranges from 5348.0 to 17669.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 3:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>

In [8]:
user_input = input("Please input brand: ")

if user_input.lower() in autos["make"].values:
    brand = autos[autos["make"] == user_input.lower()]

    print(f"The price range for {user_input} starts at {brand.sort_values(by="price", ascending=True).head(1)["price"].values} and goes to {brand.sort_values(by="price", ascending=False).head(1)["price"].values}.")
else:
    print(f"{user_input} is not a valid brand.")


The price range for volvo starts at [12940.] and goes to [22625.].


**T3:** Ask the user to input a brand, then print the number of cars in the dataset for that brand, and all attributes for a random sample car of that brand.
<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: mazda<br>
&nbsp;&nbsp;&nbsp;There are 17 cars of brand 'mazda' in the dataset.<br><br>
&nbsp;&nbsp;&nbsp;Here is the data for a random 'mazda' car:<br>
&nbsp;&nbsp;&nbsp;aspiration = std<br>
&nbsp;&nbsp;&nbsp;body-style = sedan<br>
&nbsp;&nbsp;&nbsp;bore = 3.03<br>
&nbsp;&nbsp;&nbsp;city-mpg = 31<br>
&nbsp;&nbsp;&nbsp;compression-ratio = 9.0<br>
&nbsp;&nbsp;&nbsp;curb-weight = 1945<br>
&nbsp;&nbsp;&nbsp;drive-wheels = fwd<br>
&nbsp;&nbsp;&nbsp;engine-location = front<br>
&nbsp;&nbsp;&nbsp;engine-size = 91<br>
&nbsp;&nbsp;&nbsp;engine-type = ohc<br>
&nbsp;&nbsp;&nbsp;fuel-system = 2bbl<br>
&nbsp;&nbsp;&nbsp;fuel-type = gas<br>
&nbsp;&nbsp;&nbsp;height = 54.1<br>
&nbsp;&nbsp;&nbsp;highway-mpg = 38<br>
&nbsp;&nbsp;&nbsp;horsepower = 68.0<br>
&nbsp;&nbsp;&nbsp;length = 166.8<br>
&nbsp;&nbsp;&nbsp;make = mazda<br>
&nbsp;&nbsp;&nbsp;normalized-losses = 113.0<br>
&nbsp;&nbsp;&nbsp;num-of-cylinders = four<br>
&nbsp;&nbsp;&nbsp;num-of-doors = four<br>
&nbsp;&nbsp;&nbsp;peak-rpm = 5000.0<br>
&nbsp;&nbsp;&nbsp;price = 6695.0<br>
&nbsp;&nbsp;&nbsp;stroke = 3.15<br>
&nbsp;&nbsp;&nbsp;symboling = 1<br>
&nbsp;&nbsp;&nbsp;wheel-base = 93.1<br>
&nbsp;&nbsp;&nbsp;width = 64.2<br>
</details>

In [32]:
user_input = input("Please input brand: ")

if user_input.lower() in autos["make"].values:
    brand = autos[autos["make"] == user_input.lower()]
    
    print(f"There are {len(brand)} {user_input}s in the dataset. Here's an example:")
    display(pd.DataFrame(brand.sample(1)))
    
else:
    print(f"{user_input} is not a valid brand.")

There are 11 volvos in the dataset. Here's an example:


aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
201      turbo      sedan  3.78        19                8.7         3049   

    drive-wheels engine-location  engine-size engine-type  ...   make  \
201          rwd           front          141         ohc  ...  volvo   

    normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
201              95.0              four          four    5300.0  19045.0   

    stroke  symboling wheel-base width  
201   3.15         -1      109.1  68.8  

[1 rows x 26 columns]

**T4:** Ask the user to input a brand, then export all cars of that brand to a .csv file with the same name as the brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volkswagen<br>
&nbsp;&nbsp;&nbsp;Exported 12 cars to 'volkswagen.csv'<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>

In [35]:
user_input = input("Please input brand for export: ")

if user_input.lower() in autos["make"].values:
    brand = autos[autos["make"] == user_input.lower()]
    
    brand.to_csv(f"../Data/e001_T4_{user_input}.csv")
    
else:
    print(f"{user_input} is not a valid brand.")